README.md
# Introduction

This command line app allows users to move and rename files using a custom scripting language. The app provides a powerful and efficient way to automate file organization tasks, saving users time and effort.

The app allows users to create and execute custom scripts that specify the files to move and rename, as well as the target directories and new file names. With a simple and intuitive syntax, users can quickly and easily perform complex file management tasks with just a few lines of code.

# Features

Here are some of the key features of the app:

Custom scripting language: The app provides a simple and powerful scripting language that allows users to specify exactly which files to move and rename, and where to move them to.

- File selection: Users can select files based on various criteria, such as file type, name pattern, size, and date modified.

- File renaming: Users can rename files using a variety of patterns and variables, such as the current date, file size, and file type.

- Target directories: Users can specify the target directories where they want to move their files to, and create new directories as needed.

- Preview mode: The app includes a preview mode that allows users to see the changes that will be made before actually executing the script, giving them the chance to make any necessary adjustments.


# Installing

You can install movy using pip:

```bash
pip install movy
```

Now run `movy --help` to test if the program works.

# Getting Started

All scripts are located in `~/.movy/scripts/` and each script is run in a single folder, specified by the `root`. In the example below the root is `~/Downloads`

```movy
---
root: ~/Downloads
---

[[Text Files]]
extension: txt,md
move -> ~/Documents/

[[School Stuff]]
filecontent: /school/i
move -> ~/Documents/School
```

In this example, the first block selects all files with `.txt` or `.md` extension and moves them to the `~/Documents` directory. The `~` symbol represents the home directory of the current user.

The second block selects all files that have the word "school" in their name, using a regular expression `/school/i`, and moves them to `~/Documents/School`.

You can modify this script to suit your specific file organization needs, using the custom scripting language provided by the app.

# Rules

Rules are written as `rule_name: content` and has the function to filter incoming files based on specific criteria. The available rules are listed below

## basename

this rule compare only the basename of the file (excluding the extension) with the content field. It accepts a unix glob or a regexp.

```
├── algorithms
│  ├── binarySearch.cpp
│  ├── insertionSort.cpp
│  ├── linearSearch.cpp
│  ├── mergeSort.cpp
│  ├── quickSort.cpp
│  └── selectionSort.cpp
├── Hashmap.cpp
├── Hashmap.h
├── HashmapChained.cpp
├── HashmapChained.h
├── Heap.cpp
├── Heap.h
├── Item.cpp
├── Item.h
├── Linkedlist.cpp
├── Linkedlist.h
├── List.cpp
├── List.h
├── main.cpp
├── main.exe
├── Makefile
├── Queue.cpp
├── Queue.h
└── testDataStructures.cpp

```
```movy
[[Basename Example]]
basename: List
move -> path/to/file
```

Given the file structure and the snippet listed above, this code will move all files that have the basename equals to "List" (List.cpp and List.h)

if instead of `List` as the pattern was written `Li*`, the selected files would be Linkedlist.cpp, Linkedlist.h, List.cpp and List.h


## extension

This rule extract only the extension of the file and compares it to a glob or regexp

```movy
[[Extension Example]]
extension: cpp
move -> path/to/file
```

Using the same file structure as before, this script moves all files with the `.cpp` extension. That files being: Hashmap.cpp, Heap.cpp, Linkedlist.cpp, main.cpp, Queue.cpp, HashmapChained.cpp, Item.cpp, List.cpp, PriorityList.cpp and testDataStructures.cpp

## file

This rule compares the whole path to the content field.
Using the previous examples, if the full path to that folder were `~/Documents/Codes/CPP/DataStructures` a rule:

```movy
[[File Example]]
file: /CPP/
move -> path/to/file
```

would match all of them. 

> In this example `/CPP/` is a regexp that matches CPP (case sensitive) everywhere in the string

## pdf_template

This rule use a template pdf file to compare other similar looking pdfs.

```movy
[[Pdf Template Example]]
pdf_template: {
    base_file: ~/Documents/Templates/water_bill.pdf
    score: 70
}
move -> path/to/file
```

In the snippet above, the rule will match all files that are visually similar. It will compare only the first page and the `score` arguments can be adjusted to fine tune how much similar do you want the matched documents be


## hasproperty

As files are going from rule to rule they can be assigned properties, such as regexp group contents and others


```movy
[[Has Property Example]]
basename: *rc
not hasproperty: is_syslink
move -> ~/Documents/dotfiles_backup
syslink -> 
```

This snippet:
- selects all files the ends with "rc" (such as bashrc or zshrc)
- checks if the file not has the property "is_syslink"
- move it to a backup folder
- syslink it back to the current folder

## filecontent

This rule compare a file contents to a glob or regexp.

```movy
[[Filecontent Example]]
basename: /water/i
filecontent: /Customer: *(?P<name>[a-zA-Z]+)/
move -> ~/Documents/{name}/water_bills
```

The example above first select all files that contains water in their basename, then executes a regexp that find the customer name, finally, move this this file to the folder water_bills inside a folder with the same name as the found by the regexp

# Actions

Actions are written as `action_name -> content` and has the function to act upon the incoming files modifying it or making changes to the file structure . The available actions are listed below

## echo

This action simply echoes the file name or a message to the screen.

```movy
[[Echo Example]]
basename: /water/i
echo ->
```

The output of the example above will be:

```
Echo: ~/Downloads/water bill 1.pdf
Echo: ~/Downloads/program.exe
Echo: ~/Downloads/some_file.txt
```

Is also possible use expressions in the content field to extract useful information of the file

```movy
[[Echo Example]]
basename: /water/i
echo -> {extension}
```

Output:

```
pdf
exe
txt
```

## move

This rule move files. Writing `makedirs` as argument will create all missing directories to the requested output path

## trash

This rule delete files. Using `confirm` as argument will create a confirmation prompt before every deletion




history.txt
[2022-08-13 13:27:51.100000] ./.git/__pycache__/main.py -> destinatio/main.py
[2022-08-13 13:27:51.100000] path/to/folder/file.txt -> new/destination/file.txt
[2022-08-13 13:28:44.100000] path/to/folder/second.txt -> new/destination/second.txt
[2022-08-13 13:28:50.100000] path/to/folder/third.txt -> new/destination/third.txt
[2022-08-21 21:45:50.308383] C:/Users/alanj/Downloads/toppng.com-smoke-particle-texture-399x385.png -> alternative/toppng.com-smoke-particle-texture-399x385.png
[2022-08-21 21:45:50.308383] C:/Users/alanj/Downloads/toppng.com-smoke-smoke-particle-417x278.png -> alternative/toppng.com-smoke-smoke-particle-417x278.png
[2022-08-21 21:51:46.178350] C:/Users/alanj/Downloads/kon_3-original.png -> destinatio/kon_3-original.png
[2022-08-21 21:51:46.178350] C:/Users/alanj/Downloads/Y3aISBlrbqx4qdXclY3KMJX9_fqJRySmHLuE2L1Cn2I.png -> destinatio/Y3aISBlrbqx4qdXclY3KMJX9_fqJRySmHLuE2L1Cn2I.png
[2022-08-21 21:51:46.178350] C:/Users/alanj/Downloads/kon_3-original-min.png -> destinatio/kon_3-original-min.png
[2022-08-21 21:51:46.178350] C:/Users/alanj/Downloads/wallpaper(1).png -> destinatio/wallpaper(1).png
[2022-08-21 21:51:46.178350] C:/Users/alanj/Downloads/wallpaper_storm.png -> destinatio/wallpaper_storm.png
[2022-08-21 21:51:46.178350] C:/Users/alanj/Downloads/wallpaper.png -> destinatio/wallpaper.png
[2022-08-21 21:51:46.178350] C:/Users/alanj/Downloads/toppng.com-smoke-particle-texture-399x385.png -> destinatio/toppng.com-smoke-particle-texture-399x385.png
[2022-08-21 21:51:46.178350] C:/Users/alanj/Downloads/toppng.com-smoke-smoke-particle-417x278.png -> destinatio/toppng.com-smoke-smoke-particle-417x278.png
[2022-08-21 21:51:46.178350] C:/Users/alanj/Downloads/PngItem_3392878.png -> destinatio/PngItem_3392878.png
[2022-08-21 21:52:40.682892] C:/Users/alanj/Downloads/LLVM-14.0.6-win32.exe -> executablesfolder/
[2022-08-21 21:52:40.682892] C:/Users/alanj/Downloads/ServerSetup-3.11.0.2483.exe -> executablesfolder/
[2022-08-21 21:52:40.682892] C:/Users/alanj/Downloads/vs_BuildTools.exe -> executablesfolder/
[2022-08-21 22:17:59.258117] C:/Users/alanj/Downloads/LLVM-14.0.6-win32.exe -> G:/Users/Alan/Documents/_Arquivos/Executaveis\LLVM-14.0.6-win32.exe
[2022-08-21 22:17:59.258117] C:/Users/alanj/Downloads/vs_BuildTools.exe -> G:/Users/Alan/Documents/_Arquivos/Executaveis\vs_BuildTools.exe
[2022-08-21 22:17:59.258117] C:/Users/alanj/Downloads/ServerSetup-3.11.0.2483.exe -> G:/Users/Alan/Documents/_Arquivos/Executaveis\ServerSetup-3.11.0.2483.exe
[2022-08-21 22:17:59.258117] C:/Users/alanj/Downloads/SparkARStudio_v143.msi -> G:/Users/Alan/Documents/_Arquivos/Executaveis\SparkARStudio_v143.msi
[2022-08-28 12:14:17.410393] C:/Users/alanj/Downloads/80933-171391-1-PB.pdf -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-28 12:14:17.410393] C:/Users/alanj/Downloads/annaisabel.pdf -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-28 12:14:17.523403] C:/Users/alanj/Downloads/DiscordSetup.exe -> G:/Users/Alan/Documents/_Arquivos/Executaveis\DiscordSetup.exe
[2022-08-28 14:30:29.111422] C:/Users/alanj/Downloads/DiscordSetup.exe -> G:/Users/Alan/Documents/_Arquivos/Executaveis
[2022-08-30 15:07:06.152022] C:/Users/alanj/Downloads/diff-pdf-win-0.5/fatura1.pdf -> G:/Users/Alan/Documents/_Arquivos/PDFs/Contas
[2022-08-30 15:07:06.152022] C:/Users/alanj/Downloads/diff-pdf-win-0.5/NF1.pdf -> G:/Users/Alan/Documents/_Arquivos/PDFs/Notas/NF1.pdf
[2022-08-30 15:07:06.152022] C:/Users/alanj/Downloads/diff-pdf-win-0.5/fatura2.pdf -> G:/Users/Alan/Documents/_Arquivos/PDFs/Contas
[2022-08-30 15:07:06.152022] C:/Users/alanj/Downloads/diff-pdf-win-0.5/NF2.pdf -> G:/Users/Alan/Documents/_Arquivos/PDFs/Notas/NF2.pdf
[2022-08-30 15:07:06.152022] C:/Users/alanj/Downloads/diff-pdf-win-0.5/out.pdf -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:07:06.545053] C:/Users/alanj/Downloads/diff-pdf-win-0.5/diff-pdf.exe -> G:/Users/Alan/Documents/_Arquivos/Executaveis
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/libiconv-2.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/libdeflate.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/libbrotlicommon.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/libffi-7.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/liblzma-5.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/libplc4.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/libintl-8.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/libjpeg-8.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/libnghttp2-14.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/libexpat-1.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/liblcms2-2.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/libssh2-1.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/libopenjp2-7.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/libzstd.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/libgraphite2.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs
[2022-08-30 15:30:06.299543] C:/Users/alanj/Downloads/diff-pdf-win-0.5/libbz2-1.dll -> G:/Users/Alan/Documents/_Arquivos/PDFs

classes.py
from typing import List, Optional, Tuple,Callable, Union
import os
from datetime import datetime
import re
from rich import print
from utils import check_ext, get_file_content, mergedicts

def parse_action(text:str, action_type:str, pattern_string : Optional[str] = None, flags=re.MULTILINE, rule = None) -> Tuple[str,bool]:
    new_text = ''
    non_text_transformers = ['template']
    action_results = {}

    def get_rule_options(options:dict, rule):

        if type(rule.action_content) == list:
            for opt_key, action_arg in zip(options.keys(), rule.action_content):
                options[opt_key] = action_arg

        children_dict = rule.get_children_as_dict()
        for key, child_dict in children_dict.items():
            if key in options:
                options[key] = child_dict['content']
        
        return options
    
    if action_type == 'regex':
        new_text = text
    elif action_type == 'extension':
        new_text = os.path.splitext(text)[1]
    elif action_type == 'filecontent':
        new_text = get_file_content(text)
    elif action_type == 'basename':
        new_text = os.path.splitext(os.path.basename(text))[0]
    elif action_type == 'template' and check_ext(text, 'pdf'):
        pattern_string = pattern_string.strip() if pattern_string else None
        if rule and (not pattern_string or re.search(pattern_string, text, flags=flags)):
            from similarity import pdf_similarity

            options = {
                'path': '',
                'threshold': 70
            }

            options = get_rule_options(options, rule)
            
            if check_ext(options['path'], 'pdf'):
                score = pdf_similarity(options['path'], text)
                # print([text, score*100, score*100 > options['threshold'], pattern_string])

                return '', score*100 > int(options['threshold']), action_results

    if pattern_string is not None and action_type not in non_text_transformers:
        text_match = re.search(pattern_string, new_text, flags = flags)
        if text_match:
            matchdict = text_match.groupdict()
            matchgroups = text_match.groups()
            
            if (len(matchdict) or len(matchgroups)) and not hasattr(action_results, action_type):
                action_results[action_type] = {'groups': {}}

            if len(matchdict):
                action_results[action_type]['groups'] = { **action_results[action_type]['groups'], **text_match.groupdict() }
            if len(matchgroups):
                action_results[action_type]['groups'] = { **action_results[action_type]['groups'], **dict(zip(range(1,len(matchgroups)+1), matchgroups)) }

            if len(action_results):
                print(action_results)
        return '', bool(text_match), action_results

    if new_text:
        return new_text, True, action_results

    return text, False, action_results


class HistoryItem:
    def __init__(self, date:Union[str, datetime], input_path : str, destination_path : str):
        self.date : datetime
        if isinstance(date, str):
            # 2022-08-13 13:45

            match_date = re.search(r'\[(.+)\]', date)
            if match_date:
                date = match_date.group(1)

            self.date = datetime.strptime(date, '%Y-%m-%d %H:%M:%S.%f')
        else:
            self.date = date

        self.input_path = input_path
        self.destination_path = destination_path

    def toString(self):
        return f'[{self.date}] {self.input_path} -> {self.destination_path}'

class FileHistory:
    def __init__(self, content_string : Optional[str]=None):
        self.history_array : List[HistoryItem]

        if content_string:
            self.history_array : List[HistoryItem] = self.parse(content_string) 
        else:
            self.history_array = []

    def parse(self, text) -> List[HistoryItem]:
        history = []
        for line in text.splitlines():
            regex_match = re.search(r'(\[.+?])\s*(.+?)\s*->\s*(.+?)(\[|$|\s+)', line)

            if regex_match:
                history_split = regex_match.groups()[:3]
            else:
                continue
            
            history.append(HistoryItem(*history_split)) 


        return history

    def previous_items(self):
        prev_items = []
        last_date = self.history_array[-1].date

        for h_item in reversed(self.history_array):
            if h_item.date != last_date:
                break
            if not os.path.isfile(h_item.destination_path):
                h_item.destination_path = os.path.join(h_item.destination_path, os.path.basename(h_item.input_path))

            if os.path.isfile(h_item.destination_path) and os.path.isfile(h_item.destination_path):
                prev_items.append(h_item)

        return [(item.destination_path, item.input_path, 'move') for item in prev_items]

    def filter(self, callback : Callable[[HistoryItem],bool]):
        self.history_array = [item for item in self.history_array if callback(item)]
    
    def has_item(self, history_item : HistoryItem, until_time : Optional[str] = None) -> bool:
        for item in self.history_array:
            if until_time and datetime.strptime(until_time, '%Y-%m-%d %H:%M:%S.%f') > item.date:
                break
            if history_item.input_path == item.input_path and history_item.destination_path == item.destination_path:
                return True
        return False
            
    def pop(self, n=1):
        for _ in range(n):
            self.history_array.pop()

    def remove(self, input_path, destination_path):
        for item in reversed(self.history_array):
            if item.input_path == input_path and item.destination_path == destination_path:
                self.history_array.remove(item)
                return True

        return False

    def append(self, date, input_path, destination_path):
        self.history_array.append(HistoryItem(date,input_path,destination_path))

    def extend(self, history : List[HistoryItem]):
        self.history_array.extend(history)

    def toString(self):
        output = ''

        for item in self.history_array:
            output += item.toString() + '\n'

        return output[:-1] 

class Block:
    def __init__(self, input_set, destination_set):
        self.input_set : Rule_set = input_set
        self.destination_set : Destination_set = destination_set
        self.root : None|str = None
        self.type = 'rule_set' # rule_set | routine

    def eval(self, history : Optional[FileHistory] = None):
        if self.type == 'routine':
            return []

        input_results = self.input_set.eval_children()
        destination_results = self.destination_set.eval_children(input_results)
        
        new_destination_results = []
        if history:
            for destination in destination_results:
                if not history.has_item(HistoryItem(datetime.now(), destination[0], destination[1])):
                    new_destination_results.append(destination)
                    history.append(datetime.now(), destination[0], destination[1])

        return destination_results

    def run_routine(self, blocks):
        if self.type != 'routine':
            return []

        block_names = [block.input_set.name for block in blocks]

        blocks_pending = []
        for rule in self.input_set.children:
            if rule.content.strip() not in block_names:
                continue
            current_block = blocks[block_names.index(rule.content.strip())]

            now = datetime.now()
            current_time = f'{now.hour}h{now.minute}m'

            if (type(rule.action_content) == str and current_time == rule.action_content.strip()) or rule.action_content == None: 
                blocks_pending.append(current_block)

        return blocks_pending

class Destination_set:
    def __init__(self, children) -> None:
        self.children : List[Destination_rule] = children or []

    def eval_children(self, input_rule_results: List[Tuple[str,Union[None,str]]]):
        results = []
        # input_rule_results : list[Destination_rule]


        for input_match in input_rule_results:
            command_name = 'move'

            for child in self.children:
                command_name = child.command

                # input match
                # path, name, action_action_options

                # destination_item
                # List[str,Tuple[action:str,name:str]]

                inner_output = ""
                for destination_item in child.content:
                    if isinstance(destination_item, str):
                        inner_output += destination_item
                    elif destination_item[1] is None or destination_item[1] == input_match[1]: # group name
                        dest_action,dest_name = destination_item
                        if '.' in dest_action:
                            dest_action_split = dest_action.split('.')
                            if len(dest_action_split) <= 1:
                                break
                            if dest_action_split[1] == 'groups':
                                try:
                                    if dest_action_split[2].isnumeric():
                                        inner_output += input_match[2][dest_action_split[0]]['groups'][int(dest_action_split[2])]
                                    else:
                                        inner_output += input_match[2][dest_action_split[0]]['groups'][dest_action_split[2]]
                                except:
                                    break
                        else:
                            parsed_action_text, parse_status, _ = parse_action(input_match[0], destination_item[0])
                            inner_output += parsed_action_text
                            if parse_status == False:
                                break
                    else:
                        break
                else:
                    results.append((input_match[0], inner_output, command_name))
                    break
        return results



class Input_rule:
    def __init__(self, name : Union[str, None], content : Union[str, None], action_type : Union[str, None], nested : bool):
        self.name = name or None
        self.content = content or ''
        self.action_type = action_type or 'regex'
        self.nested = nested
        self.operation = 'or'
        self.children : List[Input_rule] = []
        self.root = './'
        self.action_content = None

        time_match = re.search(r'(^\d+h$)|(^\d+h\d+m$)', self.action_type.strip())
        if time_match:
            time_split = time_match.string.split('h')
            hour = time_split[0]
            minute = time_split[1].replace('m','')or'0'
            self.action_content = f'{int(hour)}h{int(minute)}m'
            self.action_type = 'time'
        elif '|' in self.action_type:
            action_split = self.action_type.strip().split('|')

            self.action_type = action_split[0]
            self.action_content = action_split[1:]

    
    def eval(self, root:Union[str, None]=None):
        if root is None:
            root = self.root    
        regex_match = re.search(r'^\s*\/(.+)\/([a-zA-Z]+)', self.content)
        pattern_string = ''
        pattern_flags = None

        if self.action_type == 'time':
            return []

        if regex_match:
            pattern_string = regex_match.group(1)
            pattern_flags = getattr(re, regex_match.group(2)[0].upper())

            for flag_string in regex_match.group(2)[1:]:
                pattern_flags |= getattr(re, flag_string.upper())
        else:
            pattern_string = self.content
            pattern_flags = re.MULTILINE

        results = []

        for file in os.listdir(root):
            _, parse_status, action_results = parse_action(root+'/'+file, self.action_type, pattern_string, flags = pattern_flags, rule=self)
            if parse_status:

                results.append(( root+'/'+file, self.name, action_results))

        return results
    
    def get_children_as_dict(self) -> dict:
        children_dict = {}

        for i, child in enumerate(self.children):
            children_dict[child.action_type] = {
                'content': child.content,
                'action_content': child.action_content,
                'index': i
            }

        return children_dict
    
class Rule_set:
    def __init__(self, name : str, children : Union[List[Input_rule],None]) -> None:
        self.name = name
        self.children = children or []
        self.root : None|str = None

    def eval_children(self):
        final_results = []
        previous_operation = 'or'

        for child in self.children:
            current_results = child.eval(self.root)

            final_path_list = [x[0] for x in final_results]
            current_path_list = [x[0] for x in current_results]
            final_results_len = len(final_path_list)

            if previous_operation == 'or':
                # same_items = [x for x in current_results if x[0] in final_path_list]
                final_results.extend(x for x in current_results if x[0] not in final_path_list)
                for result_item in current_results:
                    path, name, action_options = result_item

                    for i in range(final_results_len):
                        final_path, final_name, final_action_options = final_results[i]

                        if final_path == path and (name is not None or final_name is None):
                            final_results[i] = (path, name, dict(mergedicts(final_action_options, action_options)))
                        else:
                            final_results[i] = (final_path, final_name, dict(mergedicts(final_action_options, action_options)))

            elif previous_operation == 'and':
                final_results = [x for x in final_results if x[0] in current_path_list]

                for i,final_item in enumerate(final_results):
                    final_path, final_name, final_action_options = final_item

                    for result_item in current_results:
                        path, name, action_options = result_item

                        if path == final_path and (name is not None or final_name is None):
                            final_results[i] = (path, name, dict(mergedicts(final_action_options, action_options)))
                        else:
                            final_results[i] = (final_path, final_name, dict(mergedicts(final_action_options, action_options)))

            previous_operation = child.operation

        print(final_results)

        return list(final_results)


class Destination_rule:
    def __init__(self, content, command : Optional[str]='move'):
        self.content : List[str|Tuple[str,str|None]] = content or []
        self.command = command

